pkg/controller/deployment/deployment_controller.go

```diff
@@ -419,17 +427,45 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
-func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Deployment) error {
-	newRC, err := dc.getNewRC(deployment)
+// Rolling back to a revision; no-op if the toRevision is deployment's current revision
+func (dc *DeploymentController) rollback(deployment extensions.Deployment, toRevision int) error {
+	if toRevision < 0 {
+		return nil
+	}
+	newRC, allOldRCs, err := dc.getNewRCAndAllOldRCs(deployment)
 	if err != nil {
 		return err
 	}
+	allRCs := append(allOldRCs, newRC)
+	// If rollback revision is 0, rollback to the last revision
+	if toRevision == 0 {
+		toRevision = secondMaxRevision(allRCs)
+	}
+	for _, rc := range allRCs {
+		v, err := revision(rc)
+		if err != nil {
+			return fmt.Errorf("unable to extract revision from deployment's RCs: %v", err)
+		}
+		glog.V(4).Infof("Found rc with revision %d: %+v", v, rc)
+		if v == toRevision {
+			glog.V(4).Infof("Found a match!")
+			// rollback by copying podTemplate.Spec from the rc, and increment revision number by 1
+			// no-op if the the spec matches current deployment's podTemplate.Spec
+			if !api.Semantic.DeepEqual(deployment.Spec.Template.Spec, rc.Spec.Template.Spec) {
```

Comment:
You are comparing twice; remove this and keep the comparison in `rollbackToTemplate`